### PR TITLE
Various improvements to `pycbc_make_skymap`

### PR DESCRIPTION
**Pull request by titodalcanton**
This PR addresses a few issues with `pycbc_make_skymap`.

First, the current implementation of `pycbc_make_skymap` tries to find the SNR peaks of above-threshold detectors on its own. However, there is no mechanism to enforce a physical light travel time between the peaks at different detectors. When the SNR peaks are not well defined, e.g. for weak triggers or in the presence of loud glitches, this can lead to nonsensical results; a typical symptom is a skymap that looks like the antenna pattern function. Hence, this PR introduces the ability to specify the exact trigger time at one or more detectors, if the time is known from a previous search (which is usually the case). Detectors for which the exact time is not specified are assumed to be subthreshold, and treated accordingly. The times are specified using the existing option `--trig-time`, like so: `--trig-time H1:123456 L1:123457`. The previous behavior is still available when using the old syntax `--trig-time 123456`.

Second, this PR fixes a bug in the code that queries the segment database to determine which detectors can be used. The code was looping over a list and at the same time modifying the list, which sometimes leads to unpredictable results.

Third, this PR introduces a `--segment-source` option to avoid unnecessary queries to GWOSC.

Fourth, the calculation of the amount of padding for the single-template processes has been tuned so as to more closely reproduce the results of the offline PyCBC search workflow. With the previous calculation, sometimes `pycbc_make_skymap` would try to use bits of data outside of science mode, or exclude usable detectors due to a trigger being very close to science mode boundaries.

Fifth: a bug prevented the skymap from being uploaded to GraceDB when using `--gracedb-server`. This is also fixed here.

Sixth: the PR adds a new example using simulated data, and adds it to the automatic inference tests.

Finally, this PR contains a massive code cleanup, including a simpler and more robust handling of the metadata associated with the various SNR and PSD series (e.g. frequency resolution and epoch). For this reason, I tried to break the cleanup into many commits, which are hopefully fairly easy to check individually if necessary.

### CODE REVIEW

**Comment by titodalcanton** [comment]
Test with GW150914: https://gracedb-playground.ligo.org/events/T411559/view/

Test with GW170817: https://gracedb-playground.ligo.org/events/T411560/view/

File: pycbc/types/frequencyseries.py [hunk 1]
```diff
@@ -430,7 +430,8 @@ def save(self, path, group=None, ifo='P1'):
                 ds = f.create_dataset(key, data=self.numpy(),
                                       compression='gzip',
                                       compression_opts=9, shuffle=True)
-                ds.attrs['epoch'] = float(self.epoch)
+                if self.epoch is not None:
+                    ds.attrs['epoch'] = float(self.epoch)
```
Comment (spxiwh) [inline pycbc/types/frequencyseries.py:434]: Is it a problem if this is *not* set. Some fiducial value would be used when reading it back in anyway. Maybe set this to that fiducial value (is it 0?) if this is None?

File: tools/pycbc_test_suite.sh [hunk 2]
```diff
@@ -58,9 +60,8 @@ if [ "$PYCBC_TEST_TYPE" = "search" ] || [ -z ${PYCBC_TEST_TYPE+x} ]; then
     fi
     popd
 
-
-    # run PyCBC Live test if running in Python 3
-    if [ "$PYTHON_VERSION" = "3" ]
+    # run PyCBC Live test if running in Python > 3.6
+    if [ "$PYTHON_VERSION" = "3" && "$PYTHON_MINOR_VERSION" -ge "7" ]
```
Comment (spxiwh) [inline tools/pycbc_test_suite.sh:64]: If this starts to get nasty, you might consider a dedicated github action for the skymap testing (in the same way that the search workflow test is a dedicated action). That would be easier to set that for specific python versions then.

File: bin/pycbc_make_skymap [hunk 3]
```diff
@@ -115,137 +125,142 @@ def main(trig_time, mass1, mass2, spin1z, spin2z, f_low, f_upper, sample_rate,
...
+    st_psd_paths = {}
+    st_out_paths = {}
+    for ifo in ifos:
+        # compose the command line for the single-template process
+        st_psd_paths[ifo] = os.path.join(
+                tmpdir, 'PSD_{}_{}.txt'.format(file_name_tag, ifo))
+        st_out_paths[ifo] = os.path.join(
+                tmpdir, 'SNRTS_{}_{}.hdf'.format(file_name_tag, ifo))
+
         command = ["pycbc_single_template",
```
Comment (spxiwh) [inline bin/pycbc_make_skymap:166]: I wonder if we should have `make_skymap` take an ini file as input to take all these options (and more gracefully handle some of the other things like segment names, injection files, data choices etc.).

Then we could use some of the workflow stuff to setup the job and then run it immediately.

More than happy for you to push back and say "Do that in a separate PR!"

**Review by spxiwh — APPROVED** [review]
See comments, but I've no objection to this going in as is, as it does improve matters a lot ... I do think the workflow code would clean this up a *lot* though (and could allow for "run at run-time" and "run in a big workflow if doing ~1000 signals" modes)

File: pycbc/types/frequencyseries.py (hunk 1 above)
Comment (titodalcanton) [reply pycbc/types/frequencyseries.py:434]: I think it *should not* be a problem in principle, i.e. I would prefer it to be set to `None` rather than to a random arbitrary time. I bet this is fine for some (most?) present applications. As far as `pycbc_make_skymap` is concerned, and for anything that runs in the automatic tests, this seems fine. I would not be surprised if this broke some edge cases, of course.

Note that when reading a `FrequencySeries` from a file, the epoch will be set to `None` if the file does not provide it (see the recent #3709).

File: bin/pycbc_make_skymap (hunk 3 above)
Comment (titodalcanton) [reply bin/pycbc_make_skymap:166]: Do that in a separate PR! :) But point taken, and I am happy to play with this for O4.

File: tools/pycbc_test_suite.sh (hunk 2 above)
Comment (titodalcanton) [reply tools/pycbc_test_suite.sh:64]: Hmm, good point. It seems the current check makes the test pass, so I am inclined to go with it (and remove it if we explicitly require ≥3.7 at some point) rather than making a special action just for this little script.
